### PR TITLE
pkg/trace/api: add warning log for out-of-cpu

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

Log a warning the APM Agent is falling behind, and point users to resource usage troubleshooting guide.

### Motivation

Make it easier for customers and support to detect out-of-cpu situations

### Additional Notes

Related to https://github.com/DataDog/documentation/pull/15126

### Describe how to test/QA your changes

- This has been tested using a custom image with [smaller channel capacity](https://github.com/DataDog/datadog-agent/blob/7.39.0-rc.6/pkg/trace/agent/agent.go#L77) (2 instead of 1000) + load tester (11k spans/s)
```
| TRACE | WARN | (run.go:261 in Warnf) | The Agent is falling behind on processing traces, 1 extra threads have been created since the Agent started. See https://docs.datadoghq.com/tracing/troubleshooting/agent_apm_resource_usage
| TRACE | WARN | (run.go:261 in Warnf) | The Agent is falling behind on processing traces, 6 extra threads have been created since the Agent started. See https://docs.datadoghq.com/tracing/troubleshooting/agent_apm_resource_usage
| TRACE | WARN | (run.go:261 in Warnf) | The Agent is falling behind on processing traces, 11 extra threads have been created since the Agent started. See https://docs.datadoghq.com/tracing/troubleshooting/agent_apm_resource_usage
| TRACE | WARN | (run.go:261 in Warnf) | The Agent is falling behind on processing traces, 16 extra threads have been created since the Agent started. See https://docs.datadoghq.com/tracing/troubleshooting/agent_apm_resource_usage
```
- Unfortunately the unit tests don't cover that code path

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/trace/api/api.go [hunk 1]
```diff
@@ -596,6 +603,17 @@ func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.
...
 }
 
+const outOfCpuLogThreshold uint32 = 5
+
+// logOutOfCPU logs a warning on the first call and every n+outOfCpuLogThreshold calls.
+// It should be called when the receiver spawns a new Goroutine to handle new requests.
+func (r *HTTPReceiver) logOutOfCPU() {
+	count := r.outOfCpuCounter.Inc()
+	if (count-1)%outOfCpuLogThreshold == 0 {
+		log.Warnf("The Agent is falling behind on processing traces, %d extra threads have been created since the Agent started. See https://docs.datadoghq.com/tracing/troubleshooting/agent_apm_resource_usage", count)
```
Comment (ahmed-mez) [inline pkg/trace/api/api.go:613]: The link depends on https://github.com/DataDog/documentation/pull/15126

File: pkg/trace/api/api.go [hunk 2]
```diff
@@ -596,6 +603,17 @@ func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.
 	}
 }
 
+const outOfCpuLogThreshold uint32 = 5
```
Comment (ahmed-mez) [inline pkg/trace/api/api.go:606]: I'm not super confident about this threshold, should we make it configurable?

**Review by maycmlee — APPROVED** [review]
👍 for docs.

File: pkg/trace/api/api.go (hunk 2 above)
Comment (ajgajg1134) [reply pkg/trace/api/api.go]: Do we have an idea on how many times this log will happen when the agent is under heavy load? Just seeing the log will be really useful for a customer so as long as we aren't accidentally helping to overwhelm the agent here by over-logging this warning (which I don't think we are?)

File: releasenotes/notes/apm-cpu-warn-4930a3d52fe06014.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---
+enhancements:
+  - |
+    The APM Agent now logs a warning when it's falling behind.
```
Comment (gbbr) [inline releasenotes/notes/apm-cpu-warn-4930a3d52fe06014.yaml:4]:
```suggestion
    APM: A warning is now logged when the agent is under heavy load.
```
We have to prefix all trace-agent entries with `APM`. Perhaps this needs to be documented somewhere. Also, I'd rephrase. "Falling behind" is ambiguous.

File: pkg/trace/api/api.go [hunk 4]
```diff
@@ -596,6 +603,17 @@ func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.
 	}
 }
 
+const outOfCPULogThreshold uint32 = 5
+
+// logOutOfCPU logs a warning on the first call and every n+outOfCPULogThreshold calls.
+// It should be called when the receiver spawns a new Goroutine to handle new requests.
+func (r *HTTPReceiver) logOutOfCPU() {
+	count := r.outOfCPUCounter.Inc()
```
Comment (gbbr) [inline pkg/trace/api/api.go:611]: IMHO this doesn't need a separate method. But I'll let you decide if you prefer it.

File: pkg/trace/api/api.go (hunk 2 above)
Comment (ahmed-mez) [reply pkg/trace/api/api.go]: Yea I'm mostly worried about it becoming spammy, at the same time we're spawning goroutines withouth limits here so we really have to let the user/support know there is an issue 😅 

> Do we have an idea on how many times this log will happen when the agent is under heavy load?
I lowered the payload channel to 50 instead off 1000 to speed-up the reproducer - I saw the warning logged 3 times in a row
```
2022-08-29 18:44:47 UTC | TRACE | WARN | (run.go:261 in Warnf) | The Agent is falling behind on processing traces, 1 extra threads have been created since the Agent started. See https://docs.datadoghq.com/tracing/troubleshooting/agent_apm_resource_usage
2022-08-29 18:44:47 UTC | TRACE | WARN | (run.go:261 in Warnf) | The Agent is falling behind on processing traces, 6 extra threads have been created since the Agent started. See https://docs.datadoghq.com/tracing/troubleshooting/agent_apm_resource_usage
2022-08-29 18:44:47 UTC | TRACE | WARN | (run.go:261 in Warnf) | The Agent is falling behind on processing traces, 11 extra threads have been created since the Agent started. See https://docs.datadoghq.com/tracing/troubleshooting/agent_apm_resource_usage
```
And then the Agent started to catch-up.
I think we'd see more warning logs in a real-life scenario (i.e with 1000 buffered payloads) so maybe we should increase `outOfCpuLogThreshold` a bit.

**Review by gbbr — APPROVED** [review]
👍🏻

File: pkg/trace/api/api.go [hunk 5]
```diff
@@ -42,6 +43,8 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/trace/watchdog"
 )
 
+const outOfCPULogThreshold uint32 = 10
```
Comment (gbbr) [inline pkg/trace/api/api.go:46]: Just curious: how / why did you pick 10? Also, you may want to document this.

Comment (ahmed-mez) [reply pkg/trace/api/api.go:46]: It's purely experimental / a tradeoff - see the discussion here https://github.com/DataDog/datadog-agent/pull/13268#discussion_r957476611
I'll add a comment in the code.

Comment (gbbr) [reply pkg/trace/api/api.go:49]: Thank you for the context. That makes complete sense. Resolving.
